crates/sui-indexer-alt/src/ingestion/mod.rs

```diff
@@ -0,0 +1,432 @@
+    /// - If a subscriber closes its channel, the ingestion service will intepret that as a signal
+    ///   to shutdown as well.
+    ///
+    /// If ingestion reaches the leading edge of the network, it will encounter checkpoints that do
+    /// not exist yet. These will be retried repeatedly on a fixed `retry_interval` until they
+    /// become available.
+    pub async fn run(self) -> Result<JoinHandle<()>> {
+        let IngestionService {
+            config,
+            client,
+            metrics,
+            subscribers,
+            cancel,
+        } = self;
+
+        if subscribers.is_empty() {
+            return Err(Error::NoSubscribers);
+        }
+
+        Ok(spawn_monitored_task!(async move {
+            let start = config.start_checkpoint;
+            info!(start, "Starting ingestion service");
+
+            match stream::iter(start..)
+                .map(Ok)
+                .try_for_each_concurrent(/* limit */ config.concurrency, |cp| {
+                    let client = client.clone();
+                    let metrics = metrics.clone();
+                    let subscribers = subscribers.clone();
+
+                    // One clone is for the supervisor to signal a cancel if it detects a
+                    // subscriber that wants to wind down ingestion, and the other is to pass to
+                    // each worker to detect cancellation.
+                    let supervisor_cancel = cancel.clone();
+                    let cancel = cancel.clone();
+
+                    // Repeatedly retry if the checkpoint is not found, assuming that we are at the
+                    // tip of the network and it will become available soon.
+                    let backoff = Constant::new(config.retry_interval);
+                    let fetch = move || {
+                        let client = client.clone();
+                        let metrics = metrics.clone();
+                        let cancel = cancel.clone();
+
+                        async move {
+                            use backoff::Error as BE;
+                            if cancel.is_cancelled() {
+                                return Err(BE::permanent(Error::Cancelled));
+                            }
+
+                            client.fetch(cp, &cancel).await.map_err(|e| match e {
+                                Error::NotFound(checkpoint) => {
+                                    debug!(checkpoint, "Checkpoint not found, retrying...");
+                                    metrics.total_ingested_not_found_retries.inc();
+                                    BE::transient(e)
+                                }
+                                e => BE::permanent(e),
+                            })
+                        }
+                    };
+
+                    async move {
+                        let checkpoint = backoff::future::retry(backoff, fetch).await?;
```

Comment:
They are retrying for different scenarios. The exponential backoff inside `fetch` is to handle transient issues (network issues, internal server error from the remote store, rate limiting, etc), constant backoff here is only to handle the "NotFound" case.

We use an exponential backoff in the first case to be a good citizen and ease pressure on the server in the hopes that it will alleviate the issue (and for some of the issues, like rate limits, it definitely will help), and we use a constant backoff here because we expect to hit this error case when we are operating at the tip of the network, and we don't want to slow down ingestion because of it in that case, so we poll at roughly the chain's checkpoint creation rate.

Reply:
Got it!